### ISSUE TITLE
Any hints on how this works?

### ISSUE THREAD

**Issue by dploeger** [mkoertgen/azure-app-exporter#6]
I've installed the Helm chart (after adding a ports section to it), provided the credentials to it and the app is running, but when I call the service /metrics only gives me internal app metrics and /apps produce a 404.

Did I miss anything or misunderstood something?

Thanks!

**Comment by mkoertgen** [comment]
Hi @dploeger ,

Thanks for reporting. Really looks like a glitch with the helm chart. I will see if i can find some time to look at it. 
Right now I need to confess that it is more of a pet project, i.e. not using it in production anywhere.

**Comment by dploeger** [comment]
Thanks!

**Comment by mkoertgen** [comment]
The endpoint should be /api/apps, btw.

**Comment by dploeger** [comment]
Ohh, okay. Will try to deploy it again later and test that.

**Comment by dploeger** [comment]
Hm. Calling that endpoint produces an exception:

``` 
INFO:     127.0.0.1:55890 - "GET /api/apps/ HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 261, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/site-packages/starlette_exporter/middleware.py", line 135, in __call__
    await self.app(scope, receive, wrapped_send)
  File "/usr/local/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 656, in __call__
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 259, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 61, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/usr/local/lib/python3.10/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/usr/local/lib/python3.10/site-packages/anyio/to_thread.py", line 28, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(func, *args, cancellable=cancellable,
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 818, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 754, in run
    result = context.run(func, *args)
  File "/app/./api/apps.py", line 20, in get_all
    return self.service.get_all()
  File "/app/./services/azure_app_service.py", line 30, in get_all
    self.observe(apps)
  File "/app/./services/azure_app_service.py", line 58, in observe
    expiry: Optional[datetime] = min(map(lambda c: c.expires, app.credentials))
ValueError: min() arg is an empty sequence
```

**Comment by dploeger** [comment]
The envs AZURE_CLIENT_ID, AZURE_CLIENT_SECRET and AZURE_TENANT_ID are set correctly.

**Comment by mkoertgen** [comment]
Probably a permission issue or an edge-case where app registrations do not have any credentials/expirations.

This will need some debugging skills. You might to check on
- [permissions](https://github.com/mkoertgen/azure-app-exporter/blob/main/_docs/service-principal.md#grant-api-permissions)
- [credentials check using azure-cli](https://github.com/mkoertgen/azure-app-exporter/blob/main/_docs/service-principal.md#finalize)

Basically, the exporter does a Microsoft Graph query like this

![image](https://github.com/mkoertgen/azure-app-exporter/assets/7235760/eaff9236-ecf3-413b-86da-cd9d037f5bb8)

**Comment by dploeger** [comment]
Ah, right. It fails on apps without credentials. See #7 for a fix.

**Comment by dploeger** [comment]
Thanks for merging. It works now.

One final question: Are the metrics updated automatically or do I need to call /api/apps/ from time to time?
